### PR TITLE
Separate out extras (breaking)

### PR DESCRIPTION
**Pull request by vinceatbluelabs**
This breaks the dependencies in setup.py into some finer-grained extras (e.g., one for each database).

It was tested by breaking the integration tests into using only the extras which ought to be required for each, and applying import laziness as needed in the code to make that work.

### CODE REVIEW

File: INSTALL.md [hunk 1]
```diff
@@ -1,3 +1,5 @@
+TODO
```
Comment (bluelabsbutler) [inline INSTALL.md:1]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/records/sources/factory.py [hunk 2]
```diff
@@ -54,7 +54,7 @@ def dataframe(self,
                   df: 'DataFrame',
                   schema_name: Optional[str]=None,
                   table_name: Optional[str]=None,
-                  db_engine: Optional[Engine]=None,
+                  db_engine: Optional['Engine']=None,
```
Comment (bluelabsbutler) [inline records_mover/records/sources/factory.py:57]: E252 missing whitespace around parameter equals

File: records_mover/records/targets/factory.py [hunk 3]
```diff
@@ -190,6 +189,8 @@ def spectrum(self,
         table, controls how any existing table found will be handled.
 
         """
+        from .spectrum import SpectrumRecordsTarget
```
Comment (bluelabsbutler) [inline records_mover/records/targets/factory.py:192]: F811 redefinition of unused 'SpectrumRecordsTarget' from line 15

File: setup.py [hunk 4]
```diff
@@ -139,48 +231,27 @@ def initialize_options(self) -> None:
...
       extras_require={
+          'airflow': airflow_dependencies,
+          'db': db_dependencies,
           'gsheets': gsheet_dependencies,
-          'movercli': gsheet_dependencies + ['typing_inspect',
-                                             'docstring_parser',
-                                             'psycopg2-binary',
-                                             'pandas<2',
-                                             'pyarrow'],
+          # TODO: or should this be cli + cli-minimal?
```
Comment (bluelabsbutler) [inline setup.py:244]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/session.py [hunk 5]
```diff
@@ -152,7 +160,14 @@ def file_url(self, url: str) -> BaseFileUrl:
     def directory_url(self, url: str) -> BaseDirectoryUrl:
         return self.url_resolver.directory_url(url)
 
-    def _boto3_session(self) -> boto3.session.Session:
+    def _boto3_session(self) -> Optional['boto3.session.Session']:
+        try:
+            import boto3
```
Comment (bluelabsbutler) [inline records_mover/session.py:165]: F811 redefinition of unused 'boto3' from line 21

File: tests/unitalldeps/test_cli_session.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+                                               mock_UrlResolver,
+                                               mock_db_driver,
+                                               mock_subprocess,
+                                               mock_os):
+        mock_subprocess.check_output.return_value = b"s3://chrisp-scratch/"
+        context = Session(session_type='cli',
+                          default_db_creds_name=None,
+                          default_aws_creds_name=None)
+        mock_subprocess.check_output.return_value = b"s3://chrisp-scratch/"
+        mock_db = Mock(name='db')
```
Comment (bluelabsbutler) [inline tests/unitalldeps/test_cli_session.py:29]: F821 undefined name 'Mock'

File: .circleci/config.yml [hunk 7]
```diff
@@ -401,8 +446,8 @@ workflows:
           name: redshift-itest
           python_version: "3.6"
           db_name: demo-itest
-          requires:
-            - test-3.8
+#          requires:  # TODO
```
Comment (bluelabsbutler) [inline .circleci/config.yml:449]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: setup.py [hunk 8]
```diff
@@ -98,6 +98,46 @@ def initialize_options(self) -> None:
         )
         self.coverage_source_file = "typecover/cobertura.xml"
 
+_itest_dependencies = [
```
Comment (bluelabsbutler) [inline setup.py:101]: E305 expected 2 blank lines after class or function definition, found 1

File: setup.py [hunk 9]
```diff
@@ -139,48 +244,28 @@ def initialize_options(self) -> None:
...
+          'airflow': airflow_dependencies,
+          'db': db_dependencies,
           'gsheets': gsheet_dependencies,
-          'movercli': gsheet_dependencies + ['typing_inspect',
-                                             'docstring_parser',
-                                             'psycopg2-binary',
-                                             'pandas<2',
-                                             'pyarrow'],
+          # TODO: or should this be cli + cli-minimal?
+          'cli': cli_dependencies_base,
```
Comment (bluelabsbutler) [inline setup.py:258]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: metrics/bigfiles_high_water_mark [hunk 10]
```diff
@@ -1 +1 @@
-875
+897
```
Comment (vinceatbluelabs) [inline metrics/bigfiles_high_water_mark:1]: setup.py (which we determined can't effectively be split out into imports) has more code now.

File: metrics/coverage_high_water_mark [hunk 11]
```diff
@@ -1 +1 @@
-93.6900
\ No newline at end of file
+93.6600
```
Comment (vinceatbluelabs) [inline metrics/coverage_high_water_mark:1]: There's a few more code twists now to do the lazy importing which were test driven by the integration tests (which don't count towards this coverage number).  Asking for two dev approval.

File: metrics/mypy_high_water_mark [hunk 12]
```diff
@@ -1 +1 @@
-89.8100
\ No newline at end of file
+89.9500
```
Comment (vinceatbluelabs) [inline metrics/mypy_high_water_mark:1]: At least the twists are well-typed!

File: records_mover/db/quoting.py [hunk 13]
```diff
@@ -1,17 +1,8 @@
...
-def _dialect(db: Optional[Union[Connection, Engine]]):
-    if db is None:
-        # use a common one which is likely to be compatible
-        return sqlalchemy_redshift.dialect.RedshiftDialect()
-    else:
-        return db.dialect
-
-
-def quote_schema_and_table(db: Optional[Union[Connection, Engine]],
+def quote_schema_and_table(db: Union[Connection, Engine],
```
Comment (vinceatbluelabs) [inline records_mover/db/quoting.py:5]: It turns out nothing passes None into this anymore anyway, so we can drop the dependency on Redshift here.

File: requirements.txt [hunk 14]
```diff
@@ -8,11 +8,9 @@ nose-progressive
 coverage
 mock
 pytz
-pandas<2
 mypy==0.770
 lxml # needed by mypy HTML coverage reporting
 pylint
-apache-airflow>=1.10,<2
```
Comment (vinceatbluelabs) [inline requirements.txt:15]: Let's install these two more explicitly through extras from `deps.sh` or `.circleci/config.yml` going forward.

File: tests/integration/records/single_db/test_records_load_df.py [hunk 15]
```diff
@@ -2,7 +2,6 @@
 import logging
 from .base_records_test import BaseRecordsIntegrationTest
 import datetime
-from odictliteral import odict
```
Comment (vinceatbluelabs) [inline tests/integration/records/single_db/test_records_load_df.py:5]: This isn't needed with the modern Python which runs the integration tests, so one less thing needed as a dependency for them.

File: metrics/bigfiles_high_water_mark [hunk 16]
```diff
@@ -1 +1 @@
-875
+896
```
Comment (vinceatbluelabs) [inline metrics/bigfiles_high_water_mark:1]: setup.py (which we determined can't effectively be split out into imports) has more code now. Asking for two dev approval.

File: Makefile [hunk 17]
```diff
@@ -21,10 +21,10 @@ citypecoverage: typecoverage
 	@test -z "$$(git status --porcelain metrics/mypy_high_water_mark)"
 
 test:
-	ENV=test nosetests --exclude='tests/integration' --cover-package=records_mover --with-coverage --with-xunit --cover-html --cover-xml --cover-inclusive
+	ENV=test nosetests --cover-package=records_mover --with-coverage --with-xunit --cover-html --cover-xml --cover-inclusive tests/unit
 
 citest:
-	ENV=test nosetests --exclude='tests/integration' --cover-package=records_mover --with-coverage --with-xunit --cover-html --cover-xml --cover-inclusive --xunit-file=test-reports/junit.xml
+	ENV=test nosetests --cover-package=records_mover --with-coverage --with-xunit --cover-html --cover-xml --cover-inclusive --xunit-file=test-reports/junit.xml tests/unit
```
Comment (vinceatbluelabs) [inline Makefile:27]: It looks like this is a more clear way to specify what we want to test.

File: metrics/coverage_high_water_mark [hunk 18]
```diff
@@ -1 +1 @@
-93.6900
\ No newline at end of file
+93.6200
```
Comment (vinceatbluelabs) [inline metrics/coverage_high_water_mark:1]: There's a few more code twists now to do the lazy importing which were test driven by the integration tests (which don't count towards this coverage number). Asking for two dev approval.

File: records_mover/records/schema/field/sqlalchemy.py [hunk 19]
```diff
@@ -47,8 +47,7 @@ def field_from_sqlalchemy_column(column: Column,
                                  type_=type_,
                                  driver=driver)
     elif (isinstance(type_, (sqlalchemy.sql.sqltypes.Numeric,
```
Comment (cwegrzyn) [inline records_mover/records/schema/field/sqlalchemy.py:49]: Hmm... Is this intentional?

File: records_mover/url/__init__.py [hunk 20]
```diff
@@ -1,25 +1,33 @@
...
+
 # These are added as modules are evaluated to avoid circular
 # dependencies
 directory_url_ctors: Dict[str, Union[Type[BaseDirectoryUrl], CreatesUrls]] = {}
 file_url_ctors: Dict[str, Union[Type[BaseFileUrl], CreatesUrls]] = {}
 
 
 def init_urls() -> None:
-    from .s3.s3_url import S3Url
+    try:
```
Comment (cwegrzyn) [inline records_mover/url/__init__.py:17]: Nice!

File: records_mover/records/schema/field/sqlalchemy.py (hunk 19 above)
Comment (vinceatbluelabs) [reply records_mover/records/schema/field/sqlalchemy.py:49]: Yep!  It turns out the postgres DOUBLE_PRECISION entry here wasn't needed, as its parent class is `sqlalchemy.sql.sqltypes.Float`

**Comment by vinceatbluelabs** [comment]
> Looks like the work here for dealing with the breaking change in existing code is adding the redshift-source / redshift-binary dep as appropriate in most places?

And for other databases, pandas, cli, etc - basically the default install for records mover is never going to be enough to do anything useful, so we'll need to adjust requirements files everywhere that records-mover is brought in.

**Pull request merged**